controllers/clustermesh/clustermesh_controller.go

```diff
@@ -123,63 +195,125 @@ func (r *ClusterMeshReconciler) reconcileDelete(ctx context.Context, cluster *cl
 
 	if len(clustermesh.Spec.Clusters) == 0 {
 		if err := r.Client.Delete(ctx, clustermesh); err != nil {
-			return ctrl.Result{}, err
+			return err
 		}
-		return ctrl.Result{}, nil
+		return nil
 	} else {
 		if err := r.Client.Update(ctx, clustermesh); err != nil {
-			return ctrl.Result{}, err
+			return err
 		}
 	}
 	return r.ReconcilePeeringsFactory(r, ctx, clustermesh)
 }
 
-func (r *ClusterMeshReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1beta1.Cluster, clustermesh *clustermeshv1beta1.ClusterMesh) (ctrl.Result, error) {
-	clSpec, err := r.PopulateClusterSpecFactory(r, ctx, cluster)
+func ReconcileSecurityGroups(r *ClusterMeshReconciler, ctx context.Context, clustermesh *clustermeshv1beta1.ClusterMesh) error {
+	// Update status
+	ownedSecurityGroupsRefs, err := crossplane.GetOwnedSecurityGroups(ctx, clustermesh, r.Client)
 	if err != nil {
-		return ctrl.Result{}, err
+		return err
 	}
+	clustermesh.Status.CrossplaneSecurityGroupRef = ownedSecurityGroupsRefs
 
-	key := client.ObjectKey{
-		Name: cluster.Labels["clusterGroup"],
+	sgPrefix := "clustermesh-"
+	cidrResults := map[string][]string{}
+	for _, cl := range clustermesh.Spec.Clusters {
+
+		// Get control plane ASG
+		kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
+		key := client.ObjectKey{
+			Namespace: "kubernetes-" + cl.Name,
+			Name:      cl.Name,
+		}
+		if err := r.Client.Get(ctx, key, kcp); err != nil {
+			return err
+		}
+		cidrResults[cl.Name] = append(cidrResults[cl.Name], kcp.Spec.KopsClusterSpec.NetworkCIDR)
 	}
 
-	if err := r.Get(ctx, key, clustermesh); err != nil {
-		if !errors.IsNotFound(err) {
-			return ctrl.Result{}, err
+	for _, cl := range clustermesh.Spec.Clusters {
+		key := client.ObjectKey{
+			Name: sgPrefix + cl.Name,
 		}
+		kopsControlPlaneSG := &sgv1beta1.SecurityGroup{}
+		if err := r.Get(ctx, key, kopsControlPlaneSG); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
 
-		ccm := crossplane.NewCrossPlaneClusterMesh(cluster.Labels["clusterGroup"], clSpec)
-		r.log.Info(fmt.Sprintf("creating clustermesh %s", ccm.ObjectMeta.GetName()))
-		if err := r.Create(ctx, ccm); err != nil {
-			return ctrl.Result{}, err
+			kopsControlPlaneSG.Name = sgPrefix + cl.Name
+			kopsControlPlaneSG.Namespace = "kubernetes-" + cl.Name
```

Comment:
This is a pattern that only we decided to use, I think we should think of an alternative way to retrieve the namespace. What about populating the clusterSpec with Namespace as well?